Initial check in for Icinga2 Provider/Resource

lrsmith opened the pull request:
Initial check in for review and comment for an Icinga2 Provider. New to GO and test all comments/critiques appreciated.

Need to add one more resource for Icinga2 Service Checks.
Need to address disabling SSL VERIFY option. disabled for now due to self signed cert.

jtopjian commented:
Icinga is something I have on my list of things to explore, so this is interesting.

Ideally, I think all of the Icinga interactions would be wrapped up into a separate Go Icinga client library, so in the Terraform provider, you'd just end up doing things like:

```go
checkResult, err := icinga.CreateCheck(...)
```

As well, all of the JSON bodies could be wrapped into some Go structs and you could have Go do most of the JSON building for you.

There are a few providers in Terraform that you can use as examples. The first one I found was [Grafana](https://github.com/hashicorp/terraform/tree/master/builtin/providers/grafana). Note how the Grafana interactions are in a separate [library](https://github.com/apparentlymart/go-grafana-api) and how that library structures the data.

Just some thoughts! I think this would be awesome to have in Terraform and if I had more time, I'd love to jump in and help 😄

stack72 commented:
Hi @lrsmith

Any thoughts on how this PR is progressing? Is this something that you are going to continue working on?

Paul

lrsmith commented:
@ stack72 

I'm refactoring based on some feedback to make it more like the grafana and rundeck providers. So I've been working on https://github.com/lrsmith/go-icinga2-api and I will modify the icinga2 provider to use that.

stack72 commented:
Very nice - look forward to trying it out. Please let me know when you need another review / testing

Paul

lrsmith commented:
@stack72 Question for you. I'm adding in a acceptance test to see if the hosts exists in the statefile https://github.com/hashicorp/terraform/pull/8306/files#diff-08e4a84d008745deef8697261646f97cR155  When I add that and run make testacc TEST=./builtin/providers/icinga2 I'm getting an error that I can figure out/get past "../../lrsmith/terraform/terraform/eval_context_builtin.go:291: cannot use vs (type map[string]"github.com/lrsmith/terraform/vendor/github.com/hashicorp/hil/ast".Variable) as type map[string]"github.com/hashicorp/terraform/vendor/github.com/hashicorp/hil/ast".Variable in argument to cfg.Interpolate
FAIL	github.com/hashicorp/terraform/builtin/providers/icinga2 [build failed]"

stack72 commented:
Hi @lrsmith 

Just checking back in on this :)

Paul

lrsmith commented:
I am finishing up the services code and once I commit that should be ready for review and comments. I'm shooting for next week to commit the code.

lrsmith commented:
Ready for initial review

stack72 commented on builtin/providers/icinga2/provider.go:
```diff
@@ -0,0 +1,81 @@
+				Type:        schema.TypeBool,
+				Optional:    true,
+				DefaultFunc: EnvBoolDefaultFunc("ICINGA2_INSECURE_SKIP_TLS_VERIFY", false),
+				Description: descriptions["insecure"],
```
I can't see "insecure" in the descriptions map?

stack72 commented on builtin/providers/icinga2/provider.go:
```diff
@@ -0,0 +1,81 @@
+		d.Get("insecure").(bool),
+	)
+
+	return config, nil
```
Are there any conditions we can actually throw an error when creating a new Client?

lrsmith replied:
Still need to look at this one.

stack72 commented on builtin/providers/icinga2/provider.go:
```diff
@@ -0,0 +1,81 @@
+func init() {
+	descriptions = map[string]string{
+		"api_url":      "The address of the Icinga2 server.\n",
+		"api_user":     "The user to authenticate to the Iccinga2 Server as.\n",
```
Typo in Iccinga

stack72 commented on builtin/providers/icinga2/provider_test.go:
```diff
@@ -0,0 +1,60 @@
+		t.Fatal("ICINGA2_API_PASSWORD must be set for acceptance tests")
+	}
+
+	v = os.Getenv("ICINGA2_INSECURE_SKIP_TLS_VERIFY")
```
Is this 100% required? It  defaults to false right?

stack72 commented on builtin/providers/icinga2/resource_icinga2_checkcommand.go:
```diff
@@ -0,0 +1,103 @@
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				//Elem:     &schema.Schema{Type: schema.TypeString},
```
Cleanup old comments please

stack72 commented on builtin/providers/icinga2/resource_icinga2_checkcommand.go:
```diff
@@ -0,0 +1,103 @@
+
+	for _, checkcommand := range checkcommands {
+		if checkcommand.Name == name {
+			d.SetId(name)
```
I think that if we don't find out command in the collection then we should error out as it means we haven't set it

lrsmith replied:
Should be addressed

stack72 replied:
I think maybe a better way to handle this would be to do something like this

```
found := false
for _, checkcommand := range checkcommands {
 		if checkcommand.Name == name {
 			d.SetId(name)
      found = true
    }
}

if !found {
  return fmt.Errorf("Failed to create Checkcommand %s : %s", name, err)
}
```

stack72 commented on builtin/providers/icinga2/resource_icinga2_checkcommand.go:
```diff
@@ -0,0 +1,103 @@
+		}
+	}
+
+	return nil
```
Create should usually call Read to set values to state

lrsmith replied:
Added sets to read, assuming create calls read? Let me know if this addresses the issues

stack72 commented on builtin/providers/icinga2/resource_icinga2_checkcommand.go:
```diff
@@ -0,0 +1,103 @@
+	return nil
+}
+
+func resourceIcinga2CheckcommandUpdate(d *schema.ResourceData, meta interface{}) error {
```
IF we don't need an Update, then we can omit this

stack72 commented on builtin/providers/icinga2/resource_icinga2_checkcommand.go:
```diff
@@ -0,0 +1,103 @@
+
+	name := d.Get("name").(string)
+
+	return client.DeleteCheckcommand(name)
```
does this not return any errors if failure?

lrsmith replied:
DeleteCheckCommand will return an error or nil, so I was just passing that back up.

stack72 commented on builtin/providers/icinga2/resource_icinga2_host.go:
```diff
@@ -0,0 +1,111 @@
+
+	for _, host := range hosts {
+		if host.Name == hostname {
+			d.SetId(hostname)
```
We should error out if we don't find our host

lrsmith replied:
Should be addressed

stack72 replied:
See CheckCommand suggestion on this

stack72 commented on builtin/providers/icinga2/resource_icinga2_host.go:
```diff
@@ -0,0 +1,111 @@
+		}
+	}
+
+	return nil
```
Create -> Read func

stack72 commented on builtin/providers/icinga2/resource_icinga2_host.go:
```diff
@@ -0,0 +1,111 @@
+	return nil
+}
+
+func resourceIcinga2HostUpdate(d *schema.ResourceData, meta interface{}) error {
```
Omit if not required

lrsmith commented:
Went through and made changes on other resource per comments

stack72 commented on builtin/providers/icinga2/resource_icinga2_hostgroup.go:
```diff
@@ -0,0 +1,95 @@
+		return err
+	}
+
+	for _, hostgroup := range hostgroups {
```
See CheckCommand for how I think this should be addressed

stack72 commented on builtin/providers/icinga2/resource_icinga2_service.go:
```diff
@@ -0,0 +1,104 @@
+		return err
+	}
+
+	for _, service := range services {
```
see CheckCommand for how I think this should be addressed

stack72 commented:
Hi @lrsmith 

I just ran the tests and got a panic. I set up the tests as follows:

```
docker run -d -ti --name icinga2-api -p 4080:80 -p 4665:5665 icinga/icinga2
export ICINGA2_API_URL=https://192.168.99.100:4665
export ICINGA2_API_USER=root
export ICINGA2_API_PASSWORD=icinga
```

Then the command `make testacc TEST=./builtin/providers/icinga2` meant this happened:

```
% make testacc TEST=./builtin/providers/icinga2                                   ✭
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/12/08 12:59:03 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/icinga2 -v  -timeout 120m
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProviderImpl
--- PASS: TestProviderImpl (0.00s)
=== RUN   TestAccCreateCheckcommand
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x1b1df6]

goroutine 208 [running]:
panic(0x4f2460, 0xc4200101c0)
	/usr/local/Cellar/go/1.7.3/libexec/src/runtime/panic.go:500 +0x1a1
github.com/hashicorp/terraform/vendor/github.com/lrsmith/go-icinga2-api/iapi.(*Server).NewAPIRequest(0xc4201de900, 0x582997, 0x3, 0xc4201a1800, 0x34, 0xc4202c88c0, 0xc3, 0x126, 0x0, 0x0, ...)
	/Users/stacko/Code/go/src/github.com/hashicorp/terraform/vendor/github.com/lrsmith/go-icinga2-api/iapi/client.go:99 +0x3e6
github.com/hashicorp/terraform/vendor/github.com/lrsmith/go-icinga2-api/iapi.(*Server).CreateCheckcommand(0xc4201de900, 0xc4200f8b21, 0x1d, 0xc4200f8b61, 0x1c, 0xc4202b3260, 0xc4201fd000, 0xc4201fd000, 0xc420186820, 0xdbdd1, ...)
	/Users/stacko/Code/go/src/github.com/hashicorp/terraform/vendor/github.com/lrsmith/go-icinga2-api/iapi/checkcommands.go:55 +0x29a
github.com/hashicorp/terraform/builtin/providers/icinga2.resourceIcinga2CheckcommandCreate(0xc42019cf00, 0x561200, 0xc4201de900, 0x800, 0xd394)
	/Users/stacko/Code/go/src/github.com/hashicorp/terraform/builtin/providers/icinga2/resource_icinga2_checkcommand.go:56 +0x32c
github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc420018de0, 0xc4201aa870, 0xc42019e640, 0x561200, 0xc4201de900, 0xc4201a1701, 0x3a, 0x0)
	/Users/stacko/Code/go/src/github.com/hashicorp/terraform/helper/schema/resource.go:162 +0x30e
github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc420018ea0, 0xc420196460, 0xc4201aa870, 0xc42019e640, 0x1, 0xc420243708, 0x9b702)
	/Users/stacko/Code/go/src/github.com/hashicorp/terraform/helper/schema/provider.go:212 +0x9b
github.com/hashicorp/terraform/terraform.(*EvalApply).Eval(0xc4201df540, 0x7d8580, 0xc42024b520, 0x2, 0x2, 0x58302d, 0x4)
	/Users/stacko/Code/go/src/github.com/hashicorp/terraform/terraform/eval_apply.go:67 +0x34b
github.com/hashicorp/terraform/terraform.EvalRaw(0x7cd8a0, 0xc4201df540, 0x7d8580, 0xc42024b520, 0x541d60, 0x0, 0x0, 0x0)
	/Users/stacko/Code/go/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x17f
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc420292b80, 0x7d8580, 0xc42024b520, 0x2, 0x2, 0x58302d, 0x4)
	/Users/stacko/Code/go/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:10 +0x6c
github.com/hashicorp/terraform/terraform.EvalRaw(0x7ce260, 0xc420292b80, 0x7d8580, 0xc42024b520, 0x7ce260, 0xc420292b80, 0xc420292ca0, 0x1a)
	/Users/stacko/Code/go/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x17f
github.com/hashicorp/terraform/terraform.Eval(0x7ce260, 0xc420292b80, 0x7d8580, 0xc42024b520, 0xc420292b80, 0x7ce260, 0xc420292b80, 0x0)
	/Users/stacko/Code/go/src/github.com/hashicorp/terraform/terraform/eval.go:34 +0x53
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x563e00, 0xc42002d998, 0x0, 0x0)
	/Users/stacko/Code/go/src/github.com/hashicorp/terraform/terraform/graph.go:301 +0xe7c
github.com/hashicorp/terraform/dag.(*AcyclicGraph).Walk.func3(0xc420265400, 0xc42023abe0, 0xc420265410, 0xc4202555f0, 0xc420265420, 0x563e00, 0xc42002d998, 0xc420239e00, 0xc420239f20)
	/Users/stacko/Code/go/src/github.com/hashicorp/terraform/dag/dag.go:246 +0x251
created by github.com/hashicorp/terraform/dag.(*AcyclicGraph).Walk
	/Users/stacko/Code/go/src/github.com/hashicorp/terraform/dag/dag.go:255 +0x6a6
exit status 2
FAIL	github.com/hashicorp/terraform/builtin/providers/icinga2	0.113s
make: *** [testacc] Error 1
```

lrsmith commented:
Paul,  thanks I will take a look

lrsmith commented:
Paul, 
I added code to verify the URL and handle the case where the server is unavailable. I also modified the code to test if the resource was found when iterating through the list.

stack72 commented:
Hi @lrsmith 

Thanks for making the changes here - do you know of a docker container I can use to test this now? When i run it against the official container, I get the following:

```
% make testacc TEST=./builtin/providers/icinga2
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/12/12 15:19:44 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/icinga2 -v  -timeout 120m
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProviderImpl
--- PASS: TestProviderImpl (0.00s)
=== RUN   TestAccCreateCheckcommand
--- FAIL: TestAccCreateCheckcommand (0.02s)
	testing.go:265: Step 0 error: Error refreshing: 1 error(s) occurred:

		* Error : Invalid API version  specified. Only v1 is currently supported.
=== RUN   TestAccCreateBasicHost
--- FAIL: TestAccCreateBasicHost (0.02s)
	testing.go:265: Step 0 error: Error refreshing: 1 error(s) occurred:

		* Error : Invalid API version  specified. Only v1 is currently supported.
=== RUN   TestAccCreateVariableHost
--- FAIL: TestAccCreateVariableHost (0.02s)
	testing.go:265: Step 0 error: Error refreshing: 1 error(s) occurred:

		* Error : Invalid API version  specified. Only v1 is currently supported.
=== RUN   TestAccCreateBasicHostGroup
--- FAIL: TestAccCreateBasicHostGroup (0.02s)
	testing.go:265: Step 0 error: Error refreshing: 1 error(s) occurred:

		* Error : Invalid API version  specified. Only v1 is currently supported.
=== RUN   TestAccCreateService
--- FAIL: TestAccCreateService (0.02s)
	testing.go:265: Step 0 error: Error refreshing: 1 error(s) occurred:

		* Error : Invalid API version  specified. Only v1 is currently supported.
FAIL
exit status 1
FAIL	github.com/hashicorp/terraform/builtin/providers/icinga2	0.108s
make: *** [testacc] Error 1
```

stack72 commented:
Ignore that comment - I should have added /v1 to the end of the URL!

This LGTM now :) 

```
% make testacc TEST=./builtin/providers/icinga2
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/12/12 15:27:32 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/icinga2 -v  -timeout 120m
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProviderImpl
--- PASS: TestProviderImpl (0.00s)
=== RUN   TestAccCreateCheckcommand
--- PASS: TestAccCreateCheckcommand (0.19s)
=== RUN   TestAccCreateBasicHost
--- PASS: TestAccCreateBasicHost (0.42s)
=== RUN   TestAccCreateVariableHost
--- PASS: TestAccCreateVariableHost (0.46s)
=== RUN   TestAccCreateBasicHostGroup
--- PASS: TestAccCreateBasicHostGroup (0.25s)
=== RUN   TestAccCreateService
--- PASS: TestAccCreateService (0.43s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/icinga2	1.760s
```

Thanks for all the work here

Paul

jtopjian commented:
Cool!

@lrsmith very nice work!
